core/src/subject/SubjectEntity.ts

```diff
@@ -49,75 +231,192 @@ export class SubjectEntity {
+      await Promise.all(
+        values.map(async ([name, value, resolve]) => {
+          // Resolve the value if necessary
+          const finalValue = resolve ? (await perspective.getExpression(value)).data : value;
+          return [name, finalValue];
+        })
+      )
+    );
+    // Assign properties to instance
+    Object.assign(instance, propsObject);
+  }
+
+  private async getData() {
+    // Builds an object with the author, timestamp, all properties, & all collections on the SubjectEntity and saves it to the instance
+    const subQueries = [buildAuthorAndTimestampQuery(), buildPropertiesQuery(), buildCollectionsQuery()];
+    const fullQuery = `
+      Base = "${this.#baseExpression}",
+      subject_class("${this.#subjectClassName}", SubjectClass),
+      ${subQueries.join(", ")}
+    `;
+
+    const result = await this.#perspective.infer(fullQuery);
+    if (result?.[0]) {
+      const { Properties, Collections, Timestamp, Author } = result?.[0];
+      const values = [...Properties, ...Collections, ["timestamp", Timestamp], ["author", Author]];
+      await SubjectEntity.assignValuesToInstance(this.#perspective, this, values);
+    }
+
+    return this;
+  }
+
+  // static async findOne(perspective: PerspectiveProxy, query?: any) {
+  //   let prologQuery = await this.getPropertiesQuery(perspective);
+  //   const result = (await perspective.infer(prologQuery))[0];
+  // }
+
+  /**
+   * Gets all instances of the subject entity in the perspective that match the query params .
+   * @param perspective - The perspective that the subject entities belongs to.
+   * @param query - The query object used to define search contraints.
+   *
+   * @returns An array of all subject entity matches.
+   *
+   */
+  static async findAll(perspective: PerspectiveProxy, query?: Query) {
+    // todo:
+    // + order
+    // + limit
+    // + offset
+    // + include
+
+    const subQueries = [
+      buildSourceQuery(query?.source),
+      buildAuthorAndTimestampQuery(query?.properties),
+      buildPropertiesQuery(query?.properties),
+      buildCollectionsQuery(query?.collections),
+      buildWhereQuery(query?.where),
+    ];
+
+    const fullQuery = `
+      findall([Base, Properties, Collections, Timestamp, Author], (
+        subject_class("${await this.getClassName(perspective)}", SubjectClass),
+        instance(SubjectClass, Base),
+        ${subQueries.filter((q) => q).join(", ")}
+      ), AllInstances)
+    `;
+
+    if (query?.where) console.log("whereQuery!!!", buildWhereQuery(query?.where));
+
+    const result = await perspective.infer(fullQuery);
+    if (!result?.[0]?.AllInstances) return [];
+
+    // Map results to instances
+    const requestedAttribtes = [...(query?.properties || []), ...(query?.collections || [])];
+    const allInstances = await Promise.all(
+      result[0].AllInstances.map(async ([Base, Properties, Collections, Timestamp, Author]) => {
+        const instance = new this(perspective, Base) as any;
```

Comment:
While this `Base` *should* really be a base of `this` Subject Class, defensive programming would be to wrap all of this in a `try {} catch()`, log errors and filter out the elements that couldn't be instantiated...